# 
# SVN commit 1031056 by winterz:
# 
# backport SVN commit 1029390 by winterz:
# 
# Fix a data corruption bug caused by cloning of recurring events.
# Now we print the first upcoming event of the recurring series, along with
# the datetime of the next one after that.
# 

Index: kontact/plugins/korganizer/apptsummarywidget.cpp
===================================================================
--- kontact/plugins/korganizer/apptsummarywidget.cpp    (révision 1031055)
+++ kontact/plugins/korganizer/apptsummarywidget.cpp    (révision 1031056)
@@ -108,6 +108,7 @@
   KIconLoader loader( "korganizer" );
   QPixmap pm = loader.loadIcon( "view-calendar-day", KIconLoader::Small );

+  QStringList uidList;
   SummaryEventInfo::setShowSpecialEvents( mShowBirthdaysFromCal,
                                           mShowAnniversariesFromCal );
   QDate currentDate = QDate::currentDate();
@@ -121,6 +122,15 @@

     foreach ( SummaryEventInfo *event, events ) {

+      Event *ev = event->ev;
+      // print the first of the recurring event series only
+      if ( ev->recurs() ) {
+        if ( uidList.contains( ev->uid() ) ) {
+          continue;
+        }
+        uidList.append( ev->uid() );
+      }
+
       // Icon label
       label = new QLabel( this );
       label->setPixmap( pm );

--- kontact/plugins/korganizer/summaryeventinfo.cpp 2009/10/03 20:55:45 1031053
+++ kontact/plugins/korganizer/summaryeventinfo.cpp 2009/10/03 21:00:11 1031056
@@ -100,30 +100,17 @@
 {
   KCal::Event *ev;
 
-  KCal::Event::List events_orig = calendar->events( date, calendar->timeSpec() );
-  KCal::Event::List::ConstIterator it = events_orig.constBegin();
+  KCal::Event::List events = calendar->events( date, calendar->timeSpec() );
+  KCal::Event::List::ConstIterator it = events.constBegin();
 
   KCal::Event::List events;
   events.setAutoDelete( true );
+
   KDateTime qdt;
   KDateTime::Spec spec = KPIM::KPimPrefs::timeSpec();
   KDateTime currentDateTime = KDateTime::currentDateTime( spec );
   QDate currentDate = currentDateTime.date();
 
-  // prevent implicitely sharing while finding recurring events
-  // replacing the QDate with the currentDate
-  for ( ; it != events_orig.constEnd(); ++it ) {
-    ev = (*it)->clone();
-    if ( ev->recursOn( date, calendar->timeSpec() ) ) {
-      qdt = ev->dtStart();
-      qdt.setDate( date );
-      ev->setDtStart( qdt );
-    }
-    if ( !skip( ev ) ) {
-      events.append( ev );
-    }
-  }
-
   // sort the events for this date by summary
   events = KCal::Calendar::sortEvents( &events,
                                        KCal::EventSortSummary,
@@ -256,6 +243,22 @@
                     KGlobal::locale()->formatTime( sET ) );
       summaryEvent->timeRange = str;
     }
+
+    // For recurring events, append the next occurrence to the time range label
+    if ( ev->recurs() ) {
+      KDateTime kdt( date, QTime( 0, 0, 0 ), KSystemTimeZones::local() );
+      kdt = kdt.addSecs( -1 );
+      KDateTime next = ev->recurrence()->getNextDateTime( kdt );
+      QString tmp = IncidenceFormatter::dateTimeToString(
+        ev->recurrence()->getNextDateTime( next ), ev->allDay(),
+        true, KSystemTimeZones::local() );
+      if ( !summaryEvent->timeRange.isEmpty() ) {
+        summaryEvent->timeRange += "<br>";
+      }
+      summaryEvent->timeRange += "<font size=\"small\"><i>" +
+                                 tmp +
+                                 "</i></font>";
+    }
   }
 
   return eventInfoList;

